token/program-2022/src/processor.rs

```diff
@@ -626,30 +630,49 @@ impl Processor {
         let dest_account_info = next_account_info(account_info_iter)?;
         let authority_info = next_account_info(account_info_iter)?;
 
-        let mut source_account = Account::unpack(&source_account_info.data.borrow())?;
-        if !source_account.is_native() && source_account.amount != 0 {
+        let source_account_data = source_account_info.data.borrow();
+        let source_account = StateWithExtensions::<Account>::unpack(&source_account_data)?;
+        if !source_account.base.is_native() && source_account.base.amount != 0 {
             return Err(TokenError::NonNativeHasBalance.into());
         }
 
         let authority = source_account
+            .base
             .close_authority
-            .unwrap_or(source_account.owner);
+            .unwrap_or(source_account.base.owner);
+
         Self::validate_owner(
             program_id,
             &authority,
             authority_info,
             account_info_iter.as_slice(),
         )?;
 
+        drop(source_account);
+        drop(source_account_data);
+
+        let mut source_account_data = source_account_info.data.borrow_mut();
+        let mut source_account =
+            StateWithExtensionsMut::<Account>::unpack(&mut source_account_data)?;
```

Comment:
All tests pass with this latest PR, and this section points out the problem: the `source_account` can't be `borrow_mut()` until after `Self::validate_owner()` is called, but needs to be `borrow()` prior.  And this workaround causes the base account to be unpacked again.

I ran into a similar issue with the initial zk-token program, which spawned [PodAccountInfoData](https://github.com/solana-labs/spl-zk-token/blob/c0b843ee9ecfb7e017e1662f5dc486c06e0723ee/program/src/pod.rs#L93-L118) that features an `into_mut()` function to promote it into an [PodAccountInfoDataMut](https://github.com/solana-labs/spl-zk-token/blob/c0b843ee9ecfb7e017e1662f5dc486c06e0723ee/program/src/pod.rs#L150-L167).  Maybe the `get_extension()` function could return a similarly wrapped struct, and `StateWithExtensionsMut::pack_base()` only borrows data mutably internally to pack the base.

Reply:
I put in #2731 to address the validate owner double-borrow annoyance.

On the other hand, an `into_mut()` function could be really useful.  I'm game to have `StateWithExtensions*` take `AccountInfo` instead.